crates/core-crnn/src/neuron.rs

```diff
@@ -0,0 +1,61 @@
+use crate::activation_function::ActivationFunction;
+use crate::thinking_layer::ThinkingLayer;
+use rand::Rng;
+use std::marker::PhantomData;
+
+pub struct Neuron<Activation: ActivationFunction> {
+    bias: f64,
+    delay: usize,
+    phantom: PhantomData<Activation>,
+}
+
+impl<Activation: ActivationFunction> Neuron<Activation> {
+    pub fn new(bias: f64, delay: usize) -> Self {
+        Self {
+            bias,
+            delay,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn random<R: Rng>(rng: &mut R) -> Self {
+        Self {
+            // TODO remove magic numbers
+            bias: rng.random::<f64>() * 2.0 - 1.0,
+            delay: rng.random_range(1..=2),
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn activate(
+        &self,
+        activate_index: usize,
+        thinking_layer: &ThinkingLayer<Activation>,
+    ) -> f64 {
+        let value: f64 = (0..thinking_layer.internal_count() - 1)
+            .map(|i| if i >= activate_index { i + 1 } else { i })
+            .map(|neuron_index| {
+                let neuron_value = thinking_layer.neuron_states()[neuron_index];
+
+                let mut weight_index =
+                    neuron_index * (thinking_layer.internal_count() - 1) + activate_index;
+                if activate_index >= neuron_index {
+                    weight_index -= 1;
+                }
+                let weight = thinking_layer.weights()[weight_index];
+
+                neuron_value * weight
+            })
+            .sum();
```

Comment:
I think iterating over all and skipping self would be a lot more readable and might even improve performance since you only need to check once for index = self instead of twice.

Reply:
Ok 👍🏻